common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.common.tracing
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import brave.Tracing
+import brave.opentracing.BraveTracer
+import brave.sampler.Sampler
+import io.opentracing.{Span, SpanContext, Tracer}
+import io.opentracing.util.GlobalTracer
+import io.opentracing.propagation.{Format, TextMapExtractAdapter, TextMapInjectAdapter}
+import zipkin2.reporter.{AsyncReporter, Sender}
+import zipkin2.reporter.okhttp3.OkHttpSender
+import pureconfig._
+import whisk.common.{LogMarkerToken, TransactionId}
+import whisk.core.ConfigKeys
+
+import scala.ref.WeakReference
+
+/**
+ * OpenTracing based implementation for tracing
+ */
+class OpenTracer(val tracer: Tracer) extends WhiskTracer {
+
+  /**
+   * Start a Trace for given service.
+   *
+   * @param transactionId transactionId to which this Trace belongs.
+   * @return TracedRequest which provides details about current service being traced.
+   */
+  override def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+
+    //initialize list for this transactionId
+    val spanList: List[WeakReference[Span]] = TracingCacheProvider.spanMap.get(transactionId.meta.id).getOrElse(Nil)
+
+    val activeSpan: Option[Span] = spanList match {
+      case Nil => {
+        val spanBuilder = tracer
+          .buildSpan(logMarker.action)
+          .withTag("transactionId", transactionId.meta.id)
+
+        Some(
+          TracingCacheProvider.contextMap
+            .get(transactionId.meta.id)
+            .map(spanBuilder.asChildOf(_).startActive(true).span())
+            .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span()))
+      }
+      case _ =>
+        Some(
+          tracer
+            .buildSpan(logMarker.action)
+            .withTag("transactionId", transactionId.meta.id)
+            .asChildOf(spanList(0).get.get)
+            .startActive(true)
+            .span())
+
+    }
+    //add active span to list
+    if (activeSpan.isDefined)
+      TracingCacheProvider.spanMap.put(transactionId.meta.id, spanList.::(new WeakReference(activeSpan.get)))
+  }
```

Comment:
Since you're returning a `Some` from each branch in the spanList match, the return type doesn't need to be an `Option`, does it?

Please note that `.get` on an Option is **not** safe, so line 69 in this case could throw an uncaught exception, if the garbage collector has collected the WeakReference.